// Copyright (c) 2011, Hauke Strasdat
// Copyright (c) 2012, Steven Lovegrove
// Copyright (c) farm-ng, inc.
//
// Licensed under the Amiga Development Kit License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://github.com/farm-ng/amiga-dev-kit/blob/main/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package farm_ng.image.proto;

message Vec2F32 {
  float x = 1;
  float y = 2;
}

// ImageSize is a message for the size of an image.
message ImageSize {
  int32 width = 1;  // width of the image.
  int32 height = 2; // height of the image.
}

// Image is a message for an image.
message Image {
  bytes data = 1; // the image data in row-major order.
  ImageSize size = 2;  // the size of the image.
  int32 num_channels = 3;  // the number of channels of the image.
  string dtype = 4;  // the data type of the image.
}

message ImageLayout {
  ImageSize size = 1;
  uint32 pitch_bytes = 2;
}

message PixelFormat {
  string number_type = 1;
  int32 num_components = 2;
  int32 num_bytes_per_component = 3;
}

message DynImage {
  ImageLayout layout = 1;
  PixelFormat pixel_format = 2;
  bytes data = 3;
}
